jax/abstract_arrays.py

```diff
@@ -101,9 +101,26 @@ def _canonicalize_dimension(dim):
   else:
     return operator.index(dim)
 
-def _canonicalize_shape(shape):
-  """Ensure shape is a tuple of int or registered _DIMENSION_TYPES."""
-  return tuple(map(_canonicalize_dimension, shape))
+def canonicalize_shape(shape):
+  """Canonicalizes and checks for errors in a user-provided shape value.
+
+  Args:
+    shape: a Python value that represents a shape.
+
+  Returns:
+    A tuple of integers.
+  """
+  try:
+    return tuple(map(_canonicalize_dimension, shape))
+  except TypeError:
+    pass
+  msg = ("Shapes must be 1D sequences of concrete values of integer type, "
+         "got {}.")
+  if any(isinstance(x, core.Tracer) and isinstance(core.get_aval(x), ShapedArray)
+         and not isinstance(core.get_aval(x), ConcreteArray) for x in shape):
+    msg += ("\nIf using `jit`, try using `static_argnums` or applying `jit` to "
+            "smaller subfunctions.")
+  raise TypeError(msg.format(shape))
```

Comment:
nit: I think it might be a little cleaner to keep this logic inside the "except" block and use the Python 3 idiom `raise ... from None` to suppress the context, e.g.,:
```python

try:
  return tuple(map(_canonicalize_dimension, shape))
except TypeError:
  msg = ("Shapes must be 1D sequences of concrete values of integer type, "
         "got {}.")
  if any(isinstance(x, core.Tracer) and isinstance(core.get_aval(x), ShapedArray)
         and not isinstance(core.get_aval(x), ConcreteArray) for x in shape):
    msg += ("\nIf using `jit`, try using `static_argnums` or applying `jit` to "
            "smaller subfunctions.")
  raise TypeError(msg.format(shape)) from None
```
```

Reply:
Eh... that style adds an indent to a whole bunch of lines and introduces a somewhat niche construct. I'll change it if you insist, but it doesn't seem worth it here.